Warn if the initial state shape doesn't match the reducer names

ellbee opened the pull request:
Ok, not super happy with this but it is a start.

* I have written tests for verifyStateShape, but I couldn't think of a neat way to test the code in createStore so there is currently a hole in the test coverage.
* I don't really like the function name verifyStateShape. Is stateShapeWarning better?
* I haven't used an external warning module. Would it be a good idea? I'm currently just calling console.warn.

Happy to take some direction on this.

gaearon commented:
Let's do this in `combineReducers` instead?

michalkvasnicak commented:
I think it's enough to put it here before `return` https://github.com/gaearon/redux/blob/breaking-changes-1.0/src/utils/combineReducers.js#L63

gaearon commented:
I think we should change `state = {}` to `state = mapValues(finalReducers, () => undefined)` so if the state isn't rehydrated, the warning is skipped.

michalkvasnicak commented:
I think that it need extra check because `typeof []` or `typeof null` return `object` too.

https://github.com/ellbee/redux/commit/0fee440b57b2c021a2b793b6423ce237c75d167f#diff-bc52e10bb986be8700c6e212e9bc581cR22

gaearon commented:
We don't need to worry about this if the check is moved to `combineReducers` because the root state is then guaranteed to be an object.

michalkvasnicak commented:
@gaearon Yeah you are right, it needs to call verifyState after mapValues are done, otherwise it'll warn everytime. I did't get mapValues right at first look.

gaearon commented:
Hmm. The initial state can still be wrong. I guess we should just check `isPlainObject` on it, shouldn't we? If it's not a plain object something is already bad.

michalkvasnicak commented:
@gaearon I agree. Initial state should be checked to be a plain object.

emmenko commented:
PS: this is a duplicate of #276 . We can close mine, but maybe you can keep the failing test?

ellbee commented:
@gaearon Yes, no problem. I've moved the check to combineReducers.

gaearon commented on src/utils/verifyStateShape.js:
```diff
@@ -0,0 +1,24 @@
+      'match shape of state object returned by reducer';
+  }
+
+  return '';
```
Better `return null` IMO.

gaearon commented on src/utils/combineReducers.js:
```diff
@@ -47,5 +50,16 @@ export default function combineReducers(reducers) {
+    if (process.env.NODE_ENV !== 'production') {
+      var warning = verifyStateShape(state, finalState);
+      /* eslint-disable no-console */
+      if (console && console.warn && warning !== '') {
```
Let's use `warning` module for consistency, `if (warning)` is also enough.

gaearon commented on src/utils/verifyStateShape.js:
```diff
@@ -0,0 +1,24 @@
+
+import isPlainObject from './isPlainObject';
+
+export default function verifyStateShape(initialState, currentState) {
```
Let's put it into `combineReducers.js`. I don't see the benefit of keeping it separately.

gaearon commented on test/utils/verifyStateShape.spec.js:
```diff
@@ -0,0 +1,13 @@
+import verifyStateShape from '../../src/utils/verifyStateShape';
+
+describe('verifyStateShape', () => {
+  it('should return warning message if object shapes differ', () => {
```
I'd rather see the unit test of `combineReducers` having correct behavior, to make sure it's actually being used correctly.

ellbee commented:
@gaearon I've hopefully addressed your comments in the latest commit

@emmenko Do you think we still need the failing test in createStore? I'll certainly add it if so.

gaearon commented on src/utils/combineReducers.js:
```diff
@@ -47,5 +56,13 @@ export default function combineReducers(reducers) {
+    warning(
+      verifyStateShape(state, finalState),
+      'Shape of initial state object given to createStore does not ' +
+      'match shape of state object returned by reducer'
```
By this point the user does not really define a reducer. All they do is pass `reducers` object (which they might get from `import * as something` ES6 import). Therefore they won't understand which reducer the warning message is referring to.

The warning should instead say something like `Unexpected key "<key>" in the initialState will be ignored. Expected to find one of the known reducer keys instead: "oneReducer", "otherReducer", ...`.

gaearon commented on test/utils/combineReducers.spec.js:
```diff
@@ -119,5 +119,17 @@ describe('Utils', () => {
+      reducer({foo: {bar: 2}});
+      reducer({bar: 2});
+      reducer(1);
+      expect(spy.calls.length).toEqual(2);
```
And let's check for a sensible warning message here using a regexp

ellbee commented:
Updated with better warning message

gaearon commented:
Looks really good. I'll test this in a day or two and get this into 1.0 release.

:+1:

gaearon commented:
Please rebase on `breaking-changes-1.0` and squash this into a single commit.

emmenko commented:
@ellbee I think it's ok like this, just it's not really the complete flow. Usually you do `createStore(reducer, initialState)` and `store.getState()`.
But I'm not sure if it's really necessary...

gaearon commented:
@emmenko Do you mean the tests? I don't think it's a problem because this is behavior completely specific to `combineReducers`.

ellbee commented:
Ok, thanks @gaearon! Rebased and squashed.

emmenko commented:
:+1:

gaearon commented on src/utils/combineReducers.js:
```diff
@@ -1,7 +1,9 @@
 import pick from '../utils/pick';
-import invariant from 'invariant';
-import { ActionTypes } from '../createStore';
+import warning from 'warning';
```
Nitpick: let's move `invariant` here so their imports are together.

ellbee replied:
Move invariant next to warning in the import order? Can do, I just put them into alphabetical order.

ellbee replied:
Ok, done

gaearon commented:
I played with this a little bit more, and I think we could still improve this a little bit:

* When the initial state is plain object, there is no warning. What I meant is, we *should always* warn if the initial state is not a plain object: that's a symptom of a mistake. There is no way for `combineReducers` to do anything useful with something other than a plain object.

* If I remove `export default` from `todos.js` in `todomvc` example, I get a cryptic error message: `Unexpected key "todos" in the initialState will be ignored. Expected to find one of the known reducer keys instead: ""`. What does `""` mean? It means `Object.keys(finalReducers).length === 0`. But it's definitely a symptom of a mistake! Evidently the user to forgot to export any reducer, so we should probably have a better warning in this situation.

* The check is being performed at every invocation, not just at the initial one. This feels like a waste even in development environment. Can we only emit the warning the first time the function is invoked? I understand it makes it impure, but it's no big deal in development environment: we just want to warn (and check) only once. Does this make sense?

emmenko commented:
Good points! 

> Can we only emit the warning the first time the function is invoked?

emmenko commented:
Sorry, my finger slipped...

I wanted to ask if it would be a problem with hot reloading?

> Can we only emit the warning the first time the function is invoked?

gaearon commented:
I don't think this would be a problem? On hot reloading the function will be re-generated so the code will run again.

emmenko commented:
Ok, just wanted to make sure :)

ellbee commented:
Updated pull request to add extra warnings and only call verifyStateShape once as [suggested](#issuecomment-124712744). Ideas for improvements to the warning messages are welcome!

I have been wondering though if these new warnings should be invariants instead?

ellbee commented:
@gaearon I'll have some spare time later today, is there anything else you would like me to do with this PR?

gaearon commented:
@ellbee 

I don't have time to look at it right now, but please do what I'd do:

* Take a sample app
* Turn your mind off and make stupid mistakes with `initialState`, exporting reducers, `export default`, etc
* See if the warnings help you or not

ellbee commented:
@gaearon OK, will do. Thanks!

gaearon commented:
Hey, awesome job. Thanks.

nuragic commented:
Hi! I just get this warning and I don't understand what's the problem with this one..

I mean, I have an initial state from the server that looks like this for example:
```js
{
  foo: [/* ... */], // ok, I've a reducer called 'foo'
  bar: true,        // no reducer
  baz: false,       // same here

}
```
Should I have one reducer for every single state property in order to access it from my components?

gaearon commented:
Yes, if you use combineReducers, because that's exactly what it does. It lets you assign a single reducer to each state property.

You can also write the root reducer by hand, without combineReducers. Then it's up to you how to structure it.

https://github.com/rackt/redux/blob/master/docs/api/combineReducers.md

https://github.com/rackt/redux/blob/master/docs/basics/Reducers.md

nuragic commented:
Thanks for the quick response! 

Ok, point taken... Maybe I need to think a bit more about my app state in order to know if I need the 1:1 relation between state properties and reducers... Another option could be to have 1 reducer to manage the entire state right?

gaearon commented:
Yes. In fact that's what combineReducers generates — a single reducer. (And instead of using combineReducers you can write it by hand.)

nuragic commented:
Ok, I'll evaluate if I need to write my own or simply use one "master" reducer (I'm doing a small app btw eheh). 

Many thanks for the support! :)
